server/repository/src/migrations/v1_02_00/context_table.rs

```diff
@@ -0,0 +1,178 @@
+use crate::migrations::sql;
+use crate::{
+    ContextRow, ContextRowRepository, ProgramFilter, ProgramRepository, ProgramRowRepository,
+    StorageConnection, PATIENT_CONTEXT_ID,
+};
+
+pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
+    sql!(
+        connection,
+        r#"
+        CREATE TABLE context (
+          id TEXT NOT NULL PRIMARY KEY,
+          name TEXT NOT NULL
+        );
+        "#
+    )?;
+
+    #[cfg(not(feature = "postgres"))]
+    sql!(
+        connection,
+        r#"
+        PRAGMA foreign_keys = 0;
+        ALTER TABLE program ADD COLUMN context_id TEXT NOT NULL DEFAULT temp REFERENCES context(id);
+        PRAGMA foreign_keys = 1;
+        "#
+    )?;
+    #[cfg(feature = "postgres")]
+    sql!(
+        connection,
+        r#"
+        ALTER TABLE program ADD COLUMN context_id TEXT NOT NULL DEFAULT '';
+        "#
+    )?;
+    // Create a context row for every existing program and update the program row
+    let programs = ProgramRepository::new(connection).query_by_filter(ProgramFilter::new())?;
+    for mut program in programs {
+        ContextRowRepository::new(connection).upsert_one(&ContextRow {
+            id: program.id.clone(),
+            name: program.name.clone(),
+        })?;
+        program.context_id = program.id.clone();
+        ProgramRowRepository::new(connection).upsert_one(&program)?;
+    }
+    #[cfg(feature = "postgres")]
+    sql!(
+        connection,
+        r#"
+        ALTER TABLE program ADD CONSTRAINT program_context_id_fkey FOREIGN KEY (context_id) REFERENCES context(id);
+        "#
+    )?;
+
+    #[cfg(not(feature = "postgres"))]
+    sql!(
+        connection,
+        r#"
+            ALTER TABLE document ADD COLUMN context_id TEXT REFERENCES context(id);
+            ALTER TABLE document DROP COLUMN context;
+
+            ALTER TABLE program_event ADD COLUMN context_id TEXT REFERENCES context(id);
+            ALTER TABLE program_event DROP COLUMN context;
+
+            ALTER TABLE document_registry ADD COLUMN context_id TEXT REFERENCES context(id);
+            ALTER TABLE document_registry DROP COLUMN document_context;
+
+            ALTER TABLE user_permission ADD COLUMN context_id TEXT REFERENCES context(id);
+            ALTER TABLE user_permission DROP COLUMN context;
+
+            ALTER TABLE program_enrolment ADD COLUMN program_id TEXT REFERENCES program(id);
+            ALTER TABLE program_enrolment DROP COLUMN context;
+
+            ALTER TABLE encounter ADD COLUMN program_id TEXT REFERENCES program(id);
+            ALTER TABLE encounter DROP COLUMN context;
+            "#
+    )?;
+
+    #[cfg(feature = "postgres")]
+    sql!(
+        connection,
+        r#"
+        ALTER TABLE document RENAME COLUMN context TO context_id;
```

Comment:
with an existing database (postgreSQL):
```
2: foreign key constraint is violated "insert or update on table \"document\" violates foreign key constraint \"document_context_id_fkey\""', /Users/m2/GitHub/open-msupply/server/server/src/lib.rs:73:10
```

and then with a new one: 
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ForeignKeyViolation("\"insert or update on table \\\"program\\\" violates foreign key constraint \\\"program_context_id_fkey\\\"\"")', /Users/m2/GitHub/open-msupply/server/server/src/lib.rs:258:10
```

Reply:
🤔  the `create_missing_master_list_and_program` is throwing the error - though I've run the upsert manually and it's ok.

Reply:
Sorry about that, I just checked the lates PR in the chain and it runs fine there.

Reply:
ah - thanks for checking. hadn't thought to check the top of the pr stack sorry